query.go

```diff
@@ -992,74 +1001,94 @@ func (s *ChainService) GetBlock(blockHash chainhash.Hash,
+	// made or not.
+	handleResp := func(req, resp wire.Message, peer string) query.Progress {
+		// The request must have been a "getdata" msg.
+		_, ok := req.(*wire.MsgGetData)
+		if !ok {
+			return noProgress
+		}
 
-				// Only set height if btcutil hasn't
-				// automagically put one in.
-				if block.Height() == btcutil.BlockHeightUnknown {
-					block.SetHeight(int32(height))
-				}
+		// We're only interested in "block" responses.
+		response, ok := resp.(*wire.MsgBlock)
+		if !ok {
+			return noProgress
+		}
 
-				// If this claims our block but doesn't pass
-				// the sanity check, the peer is trying to
-				// bamboozle us. Disconnect it.
-				if err := blockchain.CheckBlockSanity(
-					block,
-					// We don't need to check PoW because
-					// by the time we get here, it's been
-					// checked during header
-					// synchronization
-					s.chainParams.PowLimit,
-					s.timeSource,
-				); err != nil {
-					log.Warnf("Invalid block for %s "+
-						"received from %s -- "+
-						"disconnecting peer", blockHash,
-						sp.Addr())
-					sp.Disconnect()
-					return
-				}
+		// If this isn't the block we asked for, ignore it.
+		if response.BlockHash() != blockHash {
+			return noProgress
+		}
+		block := btcutil.NewBlock(response)
 
-				// TODO(roasbeef): modify CheckBlockSanity to
-				// also check witness commitment
+		// Only set height if btcutil hasn't automagically put one in.
+		if block.Height() == btcutil.BlockHeightUnknown {
+			block.SetHeight(int32(height))
+		}
 
-				// At this point, the block matches what we
-				// know about it and we declare it sane. We can
-				// kill the query and pass the response back to
-				// the caller.
-				foundBlock = block
-				close(quit)
-			default:
+		// If this claims our block but doesn't pass the sanity check,
+		// the peer is trying to bamboozle us.
+		if err := blockchain.CheckBlockSanity(
+			block,
+			// We don't need to check PoW because by the time we get
+			// here, it's been checked during header synchronization
+			s.chainParams.PowLimit,
+			s.timeSource,
+		); err != nil {
+			log.Warnf("Invalid block for %s received from %s: %v",
+				blockHash, peer, err)
+
+			// Ban and disconnect the peer.
+			err = s.BanPeer(peer, banman.InvalidBlock)
+			if err != nil {
+				log.Errorf("Unable to ban peer %v: %v", peer,
+					err)
 			}
-		},
-		options...,
-	)
+
+			return noProgress
+		}
+
+		// TODO(roasbeef): modify CheckBlockSanity to also check witness
+		// commitment
+
+		// At this point, the block matches what we know about it, and
+		// we declare it sane. We can kill the query and pass the
+		// response back to the caller.
+		foundBlock = block
+		return query.Progress{
+			Finished:   true,
+			Progressed: true,
+		}
+	}
+
+	// Prepare the query request.
+	request := &query.Request{
+		Req:        getData,
+		HandleResp: handleResp,
+	}
+
+	// Prepare the query options.
+	queryOpts := []query.QueryOption{
+		query.Encoding(qo.encoding),
```

Comment:
At a glance the option mapping here looks to be lossy, but need to dig deeper to see if it'll matter in practice (eg: we don't need things like `PeerConnectTimeout` here.

Reply:
Yeah it's defs lossy. See my overall review comment on the trickiness of the mapping between these two sets of
functional options